### PR TITLE
Support trailing zero bytes in string hash map

### LINKED ISSUE #8636: clickhouse crash " Received signal Segmentation fault"
2020.01.13 23:43:56.376109 [ 969 ] {} <Fatal> BaseDaemon: ########################################
2020.01.13 23:43:56.376119 [ 970 ] {} <Fatal> BaseDaemon: ########################################
2020.01.13 23:43:56.376170 [ 969 ] {} <Fatal> BaseDaemon: (version 19.17.6.36 (official build)) (from thread 348) Received signal Segmentation fault (11).
2020.01.13 23:43:56.376178 [ 970 ] {} <Fatal> BaseDaemon: (version 19.17.6.36 (official build)) (from thread 88) Received signal Segmentation fault (11).
2020.01.13 23:43:56.376200 [ 969 ] {} <Fatal> BaseDaemon: Address: 0x10 Access: read. Address not mapped to object.
2020.01.13 23:43:56.376200 [ 970 ] {} <Fatal> BaseDaemon: Address: 0x10 Access: read. Address not mapped to object.
2020.01.13 23:43:56.377019 [ 970 ] {} <Fatal> BaseDaemon: Stack trace: 0x5624850f3c7c 0x562485166087 0x5624850ddc82 0x5624850de5e8 0x562485079e3f 0x56248507a7ed 0x5624816cc55c 0x562487470780 0x7f43d979fe25 0x7f43d90c134d
2020.01.13 23:43:56.377022 [ 969 ] {} <Fatal> BaseDaemon: Stack trace: 0x5624850f3c7c 0x562485166087 0x5624850ddc82 0x5624850de5e8 0x562485079e3f 0x56248507a7ed 0x5624816cc55c 0x562487470780 0x7f43d979fe25 0x7f43d90c134d
2020.01.13 23:43:56.377989 [ 969 ] {} <Fatal> BaseDaemon: 3. 0x5624850f3c7c void DB::Aggregator::executeImplCase<true, DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> > >(DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&, DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >::State&, DB::Arena*, unsigned long, DB::Aggregator::AggregateFunctionInstruction*, char*) const ?
2020.01.13 23:43:56.377990 [ 970 ] {} <Fatal> BaseDaemon: 3. 0x5624850f3c7c void DB::Aggregator::executeImplCase<true, DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> > >(DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&, DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >::State&, DB::Arena*, unsigned long, DB::Aggregator::AggregateFunctionInstruction*, char*) const ?
2020.01.13 23:43:56.378571 [ 970 ] {} <Fatal> BaseDaemon: 4. 0x562485166087 void DB::Aggregator::executeImpl<DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> > >(DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&, DB::Arena*, unsigned long, std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >&, DB::Aggregator::AggregateFunctionInstruction*, bool, char*) const ?
2020.01.13 23:43:56.378572 [ 969 ] {} <Fatal> BaseDaemon: 4. 0x562485166087 void DB::Aggregator::executeImpl<DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> > >(DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&, DB::Arena*, unsigned long, std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >&, DB::Aggregator::AggregateFunctionInstruction*, bool, char*) const ?
2020.01.13 23:43:56.379415 [ 970 ] {} <Fatal> BaseDaemon: 5. 0x5624850ddc82 DB::Aggregator::executeOnBlock(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >, unsigned long, DB::AggregatedDataVariants&, std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >&, std::vector<std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >, std::allocator<std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> > > >&, bool&) ?
2020.01.13 23:43:56.379420 [ 969 ] {} <Fatal> BaseDaemon: 5. 0x5624850ddc82 DB::Aggregator::executeOnBlock(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >, unsigned long, DB::AggregatedDataVariants&, std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >&, std::vector<std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >, std::allocator<std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> > > >&, bool&) ?
2020.01.13 23:43:56.379442 [ 970 ] {} <Fatal> BaseDaemon: 6. 0x5624850de5e8 DB::Aggregator::executeOnBlock(DB::Block const&, DB::AggregatedDataVariants&, std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >&, std::vector<std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >, std::allocator<std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> > > >&, bool&) ?
2020.01.13 23:43:56.379454 [ 969 ] {} <Fatal> BaseDaemon: 6. 0x5624850de5e8 DB::Aggregator::executeOnBlock(DB::Block const&, DB::AggregatedDataVariants&, std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >&, std::vector<std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> >, std::allocator<std::vector<DB::IColumn const*, std::allocator<DB::IColumn const*> > > >&, bool&) ?
2020.01.13 23:43:56.379457 [ 970 ] {} <Fatal> BaseDaemon: 7. 0x562485079e3f DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>::thread(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long) ?
2020.01.13 23:43:56.379483 [ 969 ] {} <Fatal> BaseDaemon: 7. 0x562485079e3f DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>::thread(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long) ?
2020.01.13 23:43:56.379514 [ 970 ] {} <Fatal> BaseDaemon: 8. 0x56248507a7ed ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>::*)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>*, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long&>(void (DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>::*&&)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>*&&, std::shared_ptr<DB::ThreadGroupStatus>&&, unsigned long&)::{lambda()#1}::operator()() const ?
2020.01.13 23:43:56.379513 [ 969 ] {} <Fatal> BaseDaemon: 8. 0x56248507a7ed ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>::*)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>*, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long&>(void (DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>::*&&)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), DB::ParallelInputsProcessor<DB::ParallelAggregatingBlockInputStream::Handler>*&&, std::shared_ptr<DB::ThreadGroupStatus>&&, unsigned long&)::{lambda()#1}::operator()() const ?
2020.01.13 23:43:56.379552 [ 970 ] {} <Fatal> BaseDaemon: 9. 0x5624816cc55c ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>) ?
2020.01.13 23:43:56.379554 [ 969 ] {} <Fatal> BaseDaemon: 9. 0x5624816cc55c ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>) ?
2020.01.13 23:43:56.379572 [ 970 ] {} <Fatal> BaseDaemon: 10. 0x562487470780 ? ?
2020.01.13 23:43:56.379572 [ 969 ] {} <Fatal> BaseDaemon: 10. 0x562487470780 ? ?
2020.01.13 23:43:56.379615 [ 970 ] {} <Fatal> BaseDaemon: 11. 0x7f43d979fe25 start_thread /usr/lib64/libpthread-2.17.so
2020.01.13 23:43:56.379615 [ 969 ] {} <Fatal> BaseDaemon: 11. 0x7f43d979fe25 start_thread /usr/lib64/libpthread-2.17.so
2020.01.13 23:43:56.379639 [ 970 ] {} <Fatal> BaseDaemon: 12. 0x7f43d90c134d clone /usr/lib64/libc-2.17.so
2020.01.13 23:43:56.379639 [ 969 ] {} <Fatal> BaseDaemon: 12. 0x7f43d90c134d clone /usr/lib64/libc-2.17.so

### LINKED ISSUE #8925: rpm 19.17.4.11 (altinity) -> 20.1.2.4 (yandex) = Segmentation fault
CentOS 7.6.1810.

Раньше стояла версия кликхауса 19.17.4.11 из rpm-пакета от Altinity_clickhouse/x86_64.
После обновления кликхауса из rpm-пакета от repo.yandex.ru_clickhouse_rpm_stable_x86_64
стал периодически падать (падает в полночь два дня подряд).

```
2020.01.31 00:00:06.531882 [ 118 ] {} <Fatal> BaseDaemon: ########################################
2020.01.31 00:00:06.531938 [ 118 ] {} <Fatal> BaseDaemon: (version 20.1.2.4 (official build)) (from thread 51) (query_id: 3b7f94c7-7aae-4fe6-863e-5248cfe97cc3) Received signal Segmentation fault (11).
2020.01.31 00:00:06.531960 [ 118 ] {} <Fatal> BaseDaemon: Address: 0x7fd3e767f000 Access: read. Attempted access has violated the permissions assigned to the memory area.
2020.01.31 00:00:06.531970 [ 118 ] {} <Fatal> BaseDaemon: Stack trace: 0xbc03280 0x5695ad8 0x95c1263 0x95e8b5b 0x95e8f66 0x95e94c5 0x9651341 0x9587e60 0x4fa4657 0x4fa4c84 0x4fa3b77 0x4fa212f 0x7fd4a1aeadd5 0x7fd4a1407ead
2020.01.31 00:00:06.531993 [ 118 ] {} <Fatal> BaseDaemon: 3. 0xbc03280 memcpy  in /usr/bin/clickhouse
2020.01.31 00:00:06.532015 [ 118 ] {} <Fatal> BaseDaemon: 4. 0x5695ad8 DB::ColumnString::insertData(char const*, unsigned long)  in /usr/bin/clickhouse
2020.01.31 00:00:06.532041 [ 118 ] {} <Fatal> BaseDaemon: 5. 0x95c1263 void DB::Aggregator::convertToBlockImplFinal<DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >, StringHashMap<char*, Allocator<true, true> > >(DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&, StringHashMap<char*, Allocator<true, true> >&, std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&) const  in /usr/bin/clickhouse
2020.01.31 00:00:06.532074 [ 118 ] {} <Fatal> BaseDaemon: 6. 0x95e8b5b void DB::Aggregator::convertToBlockImpl<DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >, StringHashMap<char*, Allocator<true, true> > >(DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&, StringHashMap<char*, Allocator<true, true> >&, std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, std::__1::vector<DB::PODArray<char*, 4096ul, Allocator<false, false>, 15ul, 16ul>*, std::__1::allocator<DB::PODArray<char*, 4096ul, Allocator<false, false>, 15ul, 16ul>*> >&, std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, bool) const  in /usr/bin/clickhouse
2020.01.31 00:00:06.532094 [ 118 ] {} <Fatal> BaseDaemon: 7. 0x95e8f66 DB::Block DB::Aggregator::prepareBlockAndFill<DB::Block DB::Aggregator::convertOneBucketToBlock<DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> > >(DB::AggregatedDataVariants&, DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&, bool, unsigned long) const::'lambda'(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, std::__1::vector<DB::PODArray<char*, 4096ul, Allocator<false, false>, 15ul, 16ul>*, std::__1::allocator<DB::PODArray<char*, 4096ul, Allocator<false, false>, 15ul, 16ul>*> >&, std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, bool)>(DB::AggregatedDataVariants&, bool, unsigned long, DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&&) const  in /usr/bin/clickhouse
2020.01.31 00:00:06.532113 [ 118 ] {} <Fatal> BaseDaemon: 8. 0x95e94c5 DB::Block DB::Aggregator::convertOneBucketToBlock<DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> > >(DB::AggregatedDataVariants&, DB::AggregationMethodStringNoCache<TwoLevelStringHashMap<char*, Allocator<true, true>, StringHashMap> >&, bool, unsigned long) const  in /usr/bin/clickhouse
2020.01.31 00:00:06.532126 [ 118 ] {} <Fatal> BaseDaemon: 9. 0x9651341 DB::MergingAndConvertingBlockInputStream::thread(int, std::__1::shared_ptr<DB::ThreadGroupStatus>)  in /usr/bin/clickhouse
2020.01.31 00:00:06.532139 [ 118 ] {} <Fatal> BaseDaemon: 10. 0x9587e60 std::__1::__function::__func<std::__1::__bind<void (DB::MergingAndConvertingBlockInputStream::*)(int, std::__1::shared_ptr<DB::ThreadGroupStatus>), DB::MergingAndConvertingBlockInputStream*, int&, std::__1::shared_ptr<DB::ThreadGroupStatus> >, std::__1::allocator<std::__1::__bind<void (DB::MergingAndConvertingBlockInputStream::*)(int, std::__1::shared_ptr<DB::ThreadGroupStatus>), DB::MergingAndConvertingBlockInputStream*, int&, std::__1::shared_ptr<DB::ThreadGroupStatus> > >, void ()>::operator()()  in /usr/bin/clickhouse
2020.01.31 00:00:06.532151 [ 118 ] {} <Fatal> BaseDaemon: 11. 0x4fa4657 ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>)  in /usr/bin/clickhouse
2020.01.31 00:00:06.532163 [ 118 ] {} <Fatal> BaseDaemon: 12. 0x4fa4c84 ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() const  in /usr/bin/clickhouse
2020.01.31 00:00:06.532173 [ 118 ] {} <Fatal> BaseDaemon: 13. 0x4fa3b77 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>)  in /usr/bin/clickhouse
2020.01.31 00:00:06.532194 [ 118 ] {} <Fatal> BaseDaemon: 14. 0x4fa212f ?  in /usr/bin/clickhouse
2020.01.31 00:00:06.532207 [ 118 ] {} <Fatal> BaseDaemon: 15. 0x7dd5 start_thread  in /usr/lib64/libpthread-2.17.so
2020.01.31 00:00:06.532218 [ 118 ] {} <Fatal> BaseDaemon: 16. 0xfdead clone  in /usr/lib64/libc-2.17.so

```

На втором сервере (CentOS 7.3.1611) был также обновлён кликхаус c altinity на yandex , но подобной ошибки не возникает, правда там другие бд и запросы.

### PR DESCRIPTION
**Pull request by akuzm**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix a segmentation fault that could occur in GROUP BY over string keys containing trailing zero bytes (#8636, #8925).

Fixes #8636
Fixes #8925

### CODE REVIEW

File: dbms/Common/HashTable/StringHashTable.h [hunk 1]
```diff
@@ -237,6 +241,13 @@ class StringHashTable : private boost::noncopyable
             return func(self.m0, VoidKey{}, 0);
         }
 
+        if (x.data[sz - 1] == 0)
```
Comment (amosbird) [inline dbms/Common/HashTable/StringHashTable.h:244]: hmm, this is very critical path for performance. Let's see how perftest goes.

**Comment by amosbird** [comment]
I wonder, what's the case of having `\0` in a string. Is it somehow a binary instead? Perhaps we need a blob type

**Comment by alexey-milovidov** [comment]
@amosbird zero byte inside a string is perfectly legal.
We allow to store arbitrary binary data in strings and thus we don't need separate blob type.

Any issues with zero bytes are usually attributed to ignorant practices in old-fashioned C code.

File: dbms/Common/HashTable/StringHashTable.h (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/Common/HashTable/StringHashTable.h:244]: The branch is easy predictable and there is a hope that the memory load and arithmetic will be executed with further operations in instruction parallel manner... But I'm not sure :)

**Comment by alexey-milovidov** [comment]
There is no visible difference in performance test.

I still don't trust the new performance test for 100% (only by 99%).
@amosbird can you please also check manually with `clickhouse-benchmark`?

**Review by alexey-milovidov — APPROVED** [review]

File: dbms/Common/HashTable/StringHashMap.h [hunk 2]
```diff
@@ -25,9 +25,13 @@ struct StringHashMapCell<StringKey16, TMapped> : public HashMapCell<StringKey16,
     using Base::Base;
     static constexpr bool need_zero_value_storage = false;
     bool isZero(const HashTableNoState & state) const { return isZero(this->value.first, state); }
-    // Assuming String does not contain zero bytes. NOTE: Cannot be used in serialized method
-    static bool isZero(const StringKey16 & key, const HashTableNoState & /*state*/) { return key.low == 0; }
-    void setZero() { this->value.first.low = 0; }
+
+    // Zero means unoccupied cells in hash table. Use key with last word = 0 as
+    // zero keys, because such keys are unrepresentable (no way to encode length).
+    static bool isZero(const StringKey16 & key, const HashTableNoState &)
```
Comment (alexey-milovidov) [inline dbms/Common/HashTable/StringHashMap.h:31]: Why we only check high half?
It is because we store length in high half?

**Review by alexey-milovidov — APPROVED** [review]
Ok.

**Comment by akuzm** [comment]
I don't see a significant difference on website.xml perftest, and a couple percent slowdown on term4 in string_hash_map synthetic test (both bigger and smaller keys look OK for some reason).

File: dbms/Common/HashTable/StringHashMap.h (hunk 2 above)
Comment (akuzm) [reply dbms/Common/HashTable/StringHashMap.h:31]: yep

**Comment by amosbird** [comment]
Unfortunately I just verified with string_hash_map test locally and observed 10% regression on term48

```
term1 : 0.788     0.772
term2 : 1.527     1.521
term4 : 3.635     3.652
term8 : 7.851     7.856
term16: 9.223     9.359
term24: 10.774    10.622
term48: 13.337    14.525   <-- 10% regression
```

**Comment by alexey-milovidov** [comment]
@amosbird I see no way how to deal with it. Maybe we can find some another sort of magic.

**Pull request merged**